internal/process/manager.go

```diff
@@ -0,0 +1,448 @@
+		if !strings.Contains(ip.String(), ":") {
+			// ipv4
+			return pureIP(ip.String())
+		}
+	}
+	return ""
+}
+
+func (w *processManager) writePodInfoToCgroupMap(cgroupInode uint64, pod podInfo) error {
+	cgrinfo := cgroupInfo{}
+	if err := ebpfs.GetCgroupInfoMap().Lookup(&cgroupInode, &cgrinfo); err != nil {
+		log.Debugf("lookup cgroup info for %d error: %v", cgroupInode, err)
+	}
+	var in uint32
+	if pod.isInMesh {
+		in = 1
+	}
+	_ip, _ := linux.IP2Linux(pod.ip)
+	flag := cgrinfo.Flags
+	if pod.isAmbient {
+		flag |= AMBIENT_MESH_MODE_FLAG
+	} else {
+		flag &= ^uint16(AMBIENT_MESH_MODE_FLAG)
+	}
+	if pod.isZtunnel {
+		flag |= ZTUNNEL_FLAG
+	} else {
+		flag &= ^uint16(ZTUNNEL_FLAG)
+	}
+	tcg := cgroupInfo{
+		ID:            cgroupInode,
+		IsInMesh:      in,
+		CgroupIP:      *(*[4]uint32)(_ip),
+		Flags:         flag,
+		DetectedFlags: cgrinfo.DetectedFlags | AMBIENT_MESH_MODE_FLAG | ZTUNNEL_FLAG,
+	}
+	if reflect.DeepEqual(tcg, cgrinfo) {
+		return nil
+	}
+	return ebpfs.GetCgroupInfoMap().Update(&cgroupInode, &tcg, ebpf.UpdateAny)
+}
+
+func (w *processManager) onProcessAdded(pid uint32) error {
+	cgroup, err := getProcessCgroup(pid)
+	if err != nil {
+		log.Debugf("get process %d cgroup error: %v", pid, err)
+		return err
+	}
+	if !isInKube(cgroup) {
+		// process is not running in kubernetes
+		return nil
+	}
+	cgroupInode, err := linux.GetFileInode(path.Join(w.cgroupMountPath, cgroup))
+	if err != nil {
+		return err
+	}
+	w.lock.RLock()
+	ipkey := w.cgroupIPMap[cgroupInode]
+	w.lock.RUnlock()
+	zeroip := "0.0.0.0"
+	if ipkey == "" {
+		ips, err := getProcessIps(pid)
+		if err != nil {
+			return err
+		}
+		if !isSupportedPodProcess(ips) {
+			log.Debugf("process %d is a container process, but we not supported", pid)
+			w.lock.Lock()
+			defer w.lock.Unlock()
+			w.cgroupIPMap[cgroupInode] = zeroip
+			return nil
+		}
+		ipkey = getIPKey(ips)
+	} else if ipkey == zeroip {
+		// only for cache
+		return nil
+	}
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	w.pidCgroupMap[pid] = cgroupInode
+	w.cgroupIPMap[cgroupInode] = ipkey
+	if _, ok := w.ipCgroupsMap[ipkey]; !ok {
+		w.ipCgroupsMap[ipkey] = make(map[uint64]bool)
+	}
+	w.ipCgroupsMap[ipkey][cgroupInode] = true
+	if _, ok := w.podIPModeMap[ipkey]; !ok {
+		w.podIPModeMap[ipkey] = podInfo{
+			ip: ipkey,
+		}
+	}
+
+	return w.writePodInfoToCgroupMap(cgroupInode, w.podIPModeMap[ipkey])
+}
+
+func (w *processManager) onProcessExit(pid uint32) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	cgroupInode, ok := w.pidCgroupMap[pid]
+	if !ok {
+		return nil
+	}
+	delete(w.pidCgroupMap, pid) // delete for map
+	find := false
+	for _, cg := range w.pidCgroupMap {
+		if cgroupInode == cg {
+			find = true
+			break
+		}
+	}
+	if find {
+		// there are other processes in this cgroup, skip
+		return nil
+	}
+	// no other process in this cgroup clean up
+	log.Debugf("no process in cgroup %d, run cleanup", cgroupInode)
+	if err := ebpfs.GetCgroupInfoMap().Delete(&cgroupInode); err != nil {
```

Comment:
Given it's a LRU map, maybe we don't need to bother this?

Reply:
Although LRU can be set not to delete, it would be more reasonable to execute a deletion when we confirm it does not exist.